internal/wal/wal.go

```diff
@@ -185,5 +269,30 @@ func (w *Wal) nextSegment() error {
 	// close and fsync the current active segment
 	// create new segment file
 	// fsync the directory
+	nextId := w.active.id + 1
+	if err := w.active.Sync(); err != nil {
+		return err
+	}
+	if err := w.active.Close(); err != nil {
+		return err
+	}
+	lsn := NewLSN(nextId, int64(w.opts.MaxSegmentSize), 0)
+	seg, err := createSegment(lsn, w.opts)
+	if err != nil {
+		return err
+	}
+	w.active = seg
 	return nil
 }
+
+func (w *Wal) writeData(data []byte) (int, error) {
+	w.lock.Lock()
+	defer w.lock.Unlock()
+	pos, err := w.active.Write(data)
+	if err != nil {
+		_ = w.active.Truncate(w.active.pos)
```

Comment:
Write errors and subsequent truncate should be handled at a higher level, e.g. in wal.Write. Reason: if you successfully write part of the header into a segment, then need to open a new segment file and a write to this new file fails you need to (1) truncate the previous file, (2) delete the new file, (3) reload w.active with the previous file.